contracts/ExecutionNodeVersionBeacon.cdc

```diff
@@ -0,0 +1,425 @@
+        }
+
+        pub fun changeVersionUpdateBufferVariance(_ newUpdateBufferVariance: UFix64) {
+            ExecutionNodeVersionBeacon.versionUpdateBufferVariance = newUpdateBufferVariance
+            emit ExecutionNodeVersionUpdateBufferVarianceChanged(newVersionUpdateBufferVariance: newUpdateBufferVariance)
+
+        }
+    }
+
+    /// Returns the current updateBuffer period within which Execution Nodes
+    /// can be assured the version will not change
+    pub fun getVersionUpdateBuffer(): UInt64 {
+        return self.versionUpdateBuffer
+    }
+
+    /// Returns the current updateBuffer period within which Execution Nodes
+    /// can be assured the version will not change
+    pub fun getVersionUpdateBufferVariance(): UFix64 {
+        return self.versionUpdateBufferVariance
+    }
+
+    /// Returns a copy of the full historical versionTable
+    pub fun getVersionTable(): {UInt64: Semver} {
+        return self.versionTable
+    }
+
+    /// Function that returns that version that was defined at the most
+    /// recent block height boundary
+    pub fun getCurrentExecutionNodeVersion(): Semver? {
+        pre {
+            self.versionTable.length > 0: "No current minimum version defined."
+        }
+
+        // Update both upcomingBlockBoundaries & archivedBlockBoundaries arrays
+        self.archiveOldBlockBoundaries()
+
+        // Return nil if no historical boundaries have been defined
+        if self.archivedBlockBoundaries.length == 0 {
+            return nil
+        }
+
+        // Return the version mapped to the last historical block height boundary
+        return self.versionTable[UInt64(self.archivedBlockBoundaries.length - 1)]
+    }
+
+    /// Returns an array containing the block number at which to update and
+    /// associated version boundary - [blockBoundary: UInt64, version: Semver]
+    /// If there is no upcoming version boundary defined, returns empty array - []
+    pub fun getNextVersionBoundaryPair(): [AnyStruct] {
+
+        // Update upcomingBlockBoundaries so we know we're only dealing with future boundaries
+        self.archiveOldBlockBoundaries()
+
+        // No Future boundaries defined OR table contains no versions will
+        // return empty array as there are no boundaries to be concerned with
+        if self.upcomingBlockBoundaries.length == 0 || self.versionTable.length == 0 {
+            return []
+        }
+
+        // By now we know the next boundary we're concerned with is defined as
+        // the first element in the upcomingBlockBoundaries array
+        return [
+            self.upcomingBlockBoundaries[0],
+            self.versionTable[self.upcomingBlockBoundaries[0]]
+        ]
+    }
+
+    /// Checks whether given version was compatible at the given block height
+    pub fun isCompatibleVersion(blockHeight: UInt64, version: Semver): Bool {
+        pre {
+            blockHeight >= getCurrentBlock().height: "Given block height is in the future."
+        }
```

Comment:
This pre-condition isn't really clear. The conditional is meant to be true here but the error message looks like it is stating the thing that is supposed to be true as an error. So I would either change it to make sure that the correct condition is evaluated and error message given, or just remove this precondition entirely and return `nil` if there is an error

Reply:
Thanks for catching that. I got my pre-condition reversed. But you're right, I'll return nil instead of throwing the error in the pre-condition.

Reply:
I think it is impossible to test this because js testing doesn't support block height jumping.